### PR TITLE
Weighted losses for multitask

### PR DESCRIPTION
**Pull request by pzhanggit**
Added two types loss weights and some functions for tracking training and results visualization
Need to discuss details about better code structure

### CODE REVIEW

File: data_utils/serialized_dataset_loader.py [hunk 1]
```diff
@@ -105,7 +105,7 @@ def __update_predicted_values(self, predicted_value_option: int, data: Data):
         elif predicted_value_option == 2:
             data.y = torch.reshape(data.x[:, 1], (len(data.x), 1))
         elif predicted_value_option == 3:
-            data.y = torch.reshape(data.x[:, 1], (len(data.x), 1))
+            data.y = torch.reshape(data.x[:, 2], (len(data.x), 1))
```
Comment (streeve) [inline data_utils/serialized_dataset_loader.py:108]: I would open a separate pull request with just this change and label it a "bug"

File: examples/configuration.json [hunk 2]
```diff
@@ -5,13 +5,17 @@
     "batch_size": 64,
     "hidden_dim": 15,
     "num_conv_layers": 16,
-    "learning_rate": 0.02,
+    "learning_rate": 1e-3,
```
Comment (streeve) [inline examples/configuration.json:8]: Remove changes to this file?

File: models/PNNStack.py [hunk 3]
```diff
@@ -112,11 +140,74 @@ def loss(self, pred, value):
...
+            return self.loss_hpweighted(pred, value)
+
         pred_shape = pred.shape
         value_shape = value.shape
         if pred_shape != value_shape:
             value = torch.reshape(value, pred_shape)
-        return torch.sqrt(F.mse_loss(pred, value))
+        return torch.sqrt(F.mse_loss(pred, value)), [], []
+
+    def loss_NLL(self, pred, value):  # negative log likelihood loss
```
Comment (streeve) [inline models/PNNStack.py:155]: As we discussed, eventually this should be in the base class and used with any model (commenting to make sure there aren't any issues preventing that)

File: utils/utils.py [hunk 4]
```diff
@@ -40,75 +42,191 @@ def train_validate_test_normal(
...
+        tasklib.append(train_taskserr)
+        tasklib_vali.append(val_taskserr)
+        tasklib_test.append(test_rmse[1])
+
+        tasklib_nodes.append(train_taskserr_nodes)
+        tasklib_vali_nodes.append(val_taskserr_nodes)
+        tasklib_test_nodes.append(test_rmse[2])
+
+        ###tracking the solution evolving with training
+        # true_values=test_rmse[3]; predicted_values=test_rmse[4]
```
Comment (streeve) [inline utils/utils.py:104]: Remove?

File: utils/visualizer.py [hunk 5]
```diff
@@ -49,16 +52,468 @@ def __convert_to_list(self):
...
-            plt.savefig(
-                f"./logs/{self.model_with_config_name}/scatter_plot_test_{self.model_with_config_name}.png"
+            fig.savefig(
+                f"./logs/{self.model_with_config_name}/scatter_plot_test_ivar{ivar}.png"
             )
+            plt.close()
         else:
             plt.show()
+
+    def create_scatter_plot_atoms_hist(
```
Comment (streeve) [inline utils/visualizer.py:228]: Should be able to reuse a lot of code across plotting routines

File: models/PNNStack.py (hunk 3 above)
Comment (streeve) [inline models/PNNStack.py:155]: Is there a clean way to call the different loss functions interchangeably? Also, for each added functionality, we should add a test (at least that it runs without error)

**Review by streeve — CHANGES_REQUESTED** [review]
These are a lot of helpful changes! I really only reviewed at the surface level - I'll take a look at the details of the loss functions/ visualization when it's cleaned up a little

**Review by allaffa — APPROVED** [review]

**Comment by streeve** [comment]
Closing in favor of #31
